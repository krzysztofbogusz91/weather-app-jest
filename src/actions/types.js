export const FETCH_WEATHER = 'FETCH_WEATHER';
export const FETCH_FORECAST = 'FETCH_FORECAST';
export const USER_CORDS = 'USER_CORDS';
export const CLEAR = 'CLEAR';
export const FETCH_ERR = 'FETCH_ERR';
export const FETCH_PENDING = 'FETCH_PENDING';
export const AUTO_COM = 'AUTO_COM';
export const INPUT_VAL= 'INPUT_VAL';

