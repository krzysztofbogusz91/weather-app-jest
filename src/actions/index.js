import * as types from './types';
import { term } from '../helpers/check_search';

const API_KEY = '54df0301d1505a0aee49fe3b417ecd92';
const urlMain = `https://api.openweathermap.org/data/2.5/`;

export const clearFetch = () => {
    return {
        type: types.CLEAR,
    }
}

export const pendingFetch = () => {
    return {
        type: types.FETCH_PENDING,
        isLoading: true
    }
}

export const updateInput = term =>{
    return {
        type: types.INPUT_VAL,
        payload: term
    }
}

export const fetchWeather = data => dispatch => {

    let search = term(data);
    const url = `${urlMain}weather?${search}&appid=${API_KEY}`;



    return fetch(url)
        .then(res => res.json())
        .then(data => {
            dispatch({
                type: types.FETCH_WEATHER,
                payload: data,
                error: false,
                isLoading: false
            })
        }
        ).catch(err => {
            dispatch({
                type: types.FETCH_ERR,
                error: true,
                isLoading: false
            })
        });
}

export const fetchForecast = data => dispatch => {

    let search = term(data);
    const url = `${urlMain}forecast?${search}&appid=${API_KEY}`;

    return fetch(url)
        .then(res => res.json())
        .then(data => {
            dispatch({
                type: types.FETCH_FORECAST,
                payload: data,
                error: false,
                isLoading: false
            })
        }
        ).catch(err => {
            dispatch({
                type: types.FETCH_ERR,
                error: true,
                isLoading: false
            })

        });
}

export const fetchUser = () => dispatch => {

    // if (navigator.geolocation) {
    //     /* geolocation is available */
    //     console.log('geo yes')
    //     navigator.geolocation.getCurrentPosition((pos)=>{
    //         const crd = pos.coords
    //         dispatch({type: types.USER_CORDS, payload: {lat: crd.latitude, lng: crd.longitude} })
    //     });
    //   } else {
    /* geolocation IS NOT available */

    const key = 'AIzaSyB0jxnnSPREDEgoCvskYDxjJPJ2nSAzDaA'
    const url = `https://www.googleapis.com/geolocation/v1/geolocate?key=${key}`;

    return fetch(url, { method: 'POST' })
        .then(res => res.json())
        .then(data => {
            dispatch({
                type: types.USER_CORDS,
                payload: data.location
            })
        }
        ).catch(err => err);
    //   } end if else
}

export const autoComplete = term => dispatch => {
    
    console.log('term: ​', term );
    const data = [
    { city: "Katex", state: "Saki", country: "Azerbaijan" },

    { city: "Texcalyacac", state: "Mexico", country: "Mexico" },

    { city: "Texcoco", state: "Mexico", country: "Mexico" },

    { city: "Texmelucan", state: "Puebla", country: "Mexico" },

    { city: "Atexcatzingo", state: "Tlaxcala", country: "Mexico" },
]

   dispatch({
        type: types.AUTO_COM,
        payload: data
    })
    console.log('dispached')



    // console.log('wywo')

    // const key = `Yh5PIaZSUSmshH4egy13RmkXdUEHp1czYGmjsnWRfgrt90PVus`
    // const url = `https://andruxnet-world-cities-v1.p.mashape.com/?query=${term}&searchby=city`;

    // const myHeaders = new Headers({
    //     "Accept": "application/json",
    //     "X-Mashape-Key": key
    // });

    // const myInit = {
    //     method: 'GET',
    //     headers: myHeaders,
    //     mode: 'cors',
    //     cache: 'default'
    // };

    // const myRequest = new Request(url, myInit);

    // console.log('build​', );

    // return fetch(myRequest)
    //     .then(res => res.json())
    //     .then(data => {
    //         console.log(data)
    //         dispatch({
    //             type: types,
    //             payload: data
    //         })
    //     }
    //     ).catch(err => err);
}

// const data = [
//     { city: "Katex", state: "Saki", country: "Azerbaijan" },

//     { city: "Texcalyacac", state: "Mexico", country: "Mexico" },

//     { city: "Texcoco", state: "Mexico", country: "Mexico" },

//     { city: "Texmelucan", state: "Puebla", country: "Mexico" },

//     { city: "Atexcatzingo", state: "Tlaxcala", country: "Mexico" },
// ]