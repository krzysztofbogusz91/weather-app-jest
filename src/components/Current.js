import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchWeather, fetchUser, fetchForecast } from '../actions/index';
import { Table } from '../containers/Table';


export class Current extends Component {
    constructor(props) {
        super(props);
        this.state = {
            search: "",
            showTable: false
        }

    }

    setAll = async () => {
        await this.props.fetchUser()
        const cord = this.props.cords;
        await this.props.fetchForecast(cord)
        await this.props.fetchWeather(cord);
        this.setState({
            showTable: this.props.weather.length > 0 ? true : false
        })

    }


    componentDidMount() {
        this.setAll()
      
        //RELOAD EVERY 10 minutes? and store it all in Local storage?
        // this.time = setTimeout(()=>{
        //     this.setAll();
        //     console.log('reload'); 
        // },2000)
    }
    //NOT SOLVING PROBLEM
    // componentWillUnmount(){
    //     this.setAll().abort()
    // }

    inputChange = (e) => {
        this.setState({
            search: e.target.value
        })

    }

    tables() {
        //reduce to avoid using [0]
       const weather = this.props.today.weather.reduce((acc,bcc)=>acc.concat(bcc))
       weather.name = this.props.today.name;
       weather.temp = this.props.today.main.temp
       weather.humidity = this.props.today.main.humidity 
       
       const forecast = this.props.weather.map(a =>{
        const weat = a.weather[0];
        weat.name = a.dt_txt;
        weat.temp = a.main.temp
        weat.humidity = a.main.humidity
        return weat 
       })
       
        return (
            <div>
                <Table type="today" table={[weather]} />
                <Table type="forecast" table={forecast}/>
            </div>
        )
    }

    render() {
        //console.log(this.props)
        //console.log(this.state.showTable)


        return (
            <div className="mt-5">
                <form action="submit ">
                    <input
                        value={this.state.search}
                        onChange={this.inputChange}
                        type="text"
                        className='search-fetch form-control' />
                </form>
                <div className='current-weather'>
                    {this.state.showTable ? this.tables() : null}
                </div>
            </div>
        );
    }
}

Current.propTypes = {
    fetchWeather: PropTypes.func.isRequired,
    fetchUser: PropTypes.func.isRequired,
    fetchForecast: PropTypes.func.isRequired,
    today: PropTypes.any,
    weather: PropTypes.any,
    cords: PropTypes.any,
};

Current.defaultProps = {
    today: [],
    weather: [],
    cords: {},
}

export default connect(state => ({ ...state }), { fetchWeather, fetchUser, fetchForecast })(Current);
