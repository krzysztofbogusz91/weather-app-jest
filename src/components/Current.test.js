import React from 'react';
import { Current } from './Current'
import set from '../enzyme_set_up';
import { shallow, mount } from 'enzyme';
import { fetchWeather, fetchUser, fetchForecast } from '../actions/index';


const mockfetchWeather = jest.fn()
const mockfetchUser = jest.fn(() => { props.cords = { cords: "cor" } })
const mockfetchForecast = jest.fn();
const props = { today: [], weather: [], cords: {}, fetchUser: mockfetchUser, fetchForecast: mockfetchForecast, fetchWeather: mockfetchWeather }

describe('Current', () => {

    //USE MOUNT TO TEST PROPS VALUES AND CHANGES
    const current = shallow(<Current {...props} />)

    it('renders without crashing', () => {
        expect(current).toMatchSnapshot();
    });

    it('should have initial empty state ', function () {
        const bar = current.find('.search-fetch');
        expect(current.state().search).toEqual("");
    });

    it('should have input search bar that sets state ', function () {
        const bar = current.find('.search-fetch');
        bar.simulate('change', { target: { value: "Lond" } })
        expect(current.state().search).toEqual("Lond");
    });

    it('should have search button', function () {
        const button = current.find('.search');
        
        expect(button.exists()).toBe(true);
    });

  

    //NOT NECESSARY => just other way than above
    it('should call inputChage method and c', () => {
        current.instance().inputChange({ target: { value: "Lod" } })
        expect(current.state().search).toEqual("Lod");
    });

    it('fetchUser should update props', () => {
        expect(current.props().cords).not.toEqual({})
    });

    it('should not show table when showTable state falase', () => {
        expect(current.find('.current-weather').children().length).toEqual(0)
    });
})

describe('Mount Current', () => {

    const current = mount(<Current {...props} />)

    it('should call fetchUser prop once', () => {
        expect(current.props().fetchUser).toHaveBeenCalled();
    });

    it('On button click should call feches with search state', function () {
        current.setState({
            search: "Lond"
        })
        const button = current.find('.search').simulate('click');
        
        expect(current.props().fetchForecast).toHaveBeenCalledWith("Lond");
        expect(current.props().fetchWeather).toHaveBeenCalledWith("Lond");
    });



    it('fetchUser should update props', () => {
        expect(current.props().cords).not.toEqual({})
    });

    it(' show table when weather length is > 0', () => {
        current.setProps({
            today: [{
                    description: "clear sky",
                    humidity: 45,
                    icon: "01d",
                    main: "Clear",
                    name: "Pozna≈Ñ",
                    temp: 294.31
                }],
            weather: [
                {
                    description: "clear sky",
                    humidity: 45,
                    icon: "01d",
                    id: 800,
                    main: "Clear",
                    name: "2018-05-07 18:00:00",
                    temp: 294.31
                },
                {
                    description: "clear sky",
                    humidity: 45,
                    icon: "01d",
                    id: 800,
                    main: "Clear",
                    name: "2018-05-07 18:00:00",
                    temp: 294.31
                },
            ]
        })

        expect(current.find('.current-weather').children().length).toEqual(1)
    });

});

