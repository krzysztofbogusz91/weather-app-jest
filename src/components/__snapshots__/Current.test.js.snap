// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Current renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Current
    cords={Object {}}
    fetchForecast={[Function]}
    fetchUser={[Function]}
    fetchWeather={[Function]}
    mockClearFetch={[Function]}
    pendingFetch={[Function]}
    today={Array []}
    weather={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="current-weather"
>
        <Wrapper
                lat={undefined}
                lng={undefined}
        />
</div>,
      "className": "mt-5",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Wrapper
            lat={undefined}
            lng={undefined}
/>,
          null,
          null,
        ],
        "className": "current-weather",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "lat": undefined,
            "lng": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
        null,
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="current-weather"
>
          <Wrapper
                    lat={undefined}
                    lng={undefined}
          />
</div>,
        "className": "mt-5",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Wrapper
              lat={undefined}
              lng={undefined}
/>,
            null,
            null,
          ],
          "className": "current-weather",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "lat": undefined,
              "lng": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          null,
          null,
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
